public/js/repeater.js

```diff
@@ -0,0 +1,185 @@
+window.Repeater = (function() {
+  "use strict"
+
+  var repeatedSets
+
+  // Since we can't depend on Array.prototype.map
+  function map(arr, fn) {
```

Comment:
Oh neat! Does that get us ES.next features as well?

Reply:
Yes.

You can use whatever you need and we'll gets things setup to ensure it compiles back.

Reply:
Okay! How about we get that set up first, so I can test it?

Reply:
If you look at the fileInput in the utils ... the fileInput gets exported

```javascript
export const fileInput = () => {
 ....
```

Than in the personal route in the js dir it's imported

```javascript
import { fileInput } from '../../../utils/fileInput'
import { registerToggle } from '../../../utils/toggleClass'
;(function(document, window, index) {
  fileInput()
  registerToggle('.multiple-choice__item input', '.notify-type-toggle-container')
})(document, window, 0)

```

In the Wepack config you need to register the file under the entry "key"

```
const config = getConfig({
    mode: argv.mode,
    entry: {
      styles: './assets/scss/app.scss',
      personal: './routes/personal/js/personal.js',
      addresses: 'your-path-here'
    },
```

In the controller if you use 
```
const js = getClientJs(req, route.name)
```

That code will take care of the path 

And you pass it to the view

```
routeUtils.getViewData(req, { jsFiles: js ? [js] : false }),
```

Reply:
If we need to target or add things to Webpack that can be done here https://github.com/cds-snc/webpack-starter

Or the config can be modified locally using Webpack merge.

Reply:
I'm *very* confused about this bit - why is the webpack stuff a separate repository? How is it configured?

Reply:
(also seeing as we're deprecating `getViewData` we'll need another way of passing that)

Reply:
> I'm _very_ confused about this bit - why is the webpack stuff a separate repository? How is it configured?

In general devs don't "enjoy" or know how to configure Webpack so we're doing it for them and allowing them them to override.  Most projects will need roughly the same config so much like the starter nice to give them a base setup.  Projects such as Vue CLI, Next JS, Razzle and a bunch of others follow this pattern.  

On top of this again like this repo we can manage the dependancies on layer up - so bump once on X packages upstream and the end user just needs to bump one package.  IMO better to do that work up a level if we can.

As far as config:  

If you look at the Webpack file in this repo

https://github.com/cds-snc/node-starter-app/blob/master/webpack.config.js

The other repo gives you some config defaults 

```
const path = require('path')

module.exports = (env, argv) => {
  const { getConfig } = require('@cdssnc/webpack-starter')
  const config = getConfig({
    ...
  return config
}
```

And webpack merge can be used to extend that config or the user can opt out and just use there own config.

Reply:
> (also seeing as we're deprecating `getViewData` we'll need another way of passing that)

Understand about deprecating `getViewData`  (or is it being removed) but not sure what the question is here.

I can lend a hand getting the Webpack stuff going if you like.

Reply:
I see! So if I understand right, the repo there is just a dummy package used to pull in dependencies.

Not a question, just adding a TODO for myself, because the way we handle `jsFiles` needs to be thought out a little more with the new way of rendering. Perhaps it can be rolled into `route.render()`.

Reply:
Right it gives a based setup and essentially stops the duplication of have all those deps in all the repos.

Also note:  In that repo I wrote a custom Webpack plugin that writes a "_filelist.json" file

Which is just a lookup file.


```_filelist.json
{
	"book":"book.930291000905c8c00779.js"
}
```

To avoid cache issues we have Wepack setup to add a hash to the end of the filename
```
 output: {
      filename: '[name].[chunkhash].js',
```

So technically for each route we don't really know what the file name will be.  We look it up using the _filelist.json.  That part all works fine just wanted to mention the reason it's there.

At some point I want to get into hot reloading js file (replacing them in page vs nodemon full page reload) but that's down the road.